File: leetcode/cpp/coinChange.py

def coinChange(coins: List[int], amount: int):

  def dp(n):
    #base case
    if n == 0: return 0
    if n < 0:  return -1
    
    #求最小值， 所以初始化为正无穷
    res = float('INF')
    for coin in coins:
      subproblem = dp(n - coin)
      #子问题无解，跳过
      if subproblem == -1: continue
      res = min(res, 1 + subproblem)

    return res if res != float('INF') else -1

  return dp(amount)

File: leetcode/cpp/coinChange.cpp

#include <vector>

using namespace std;


// dp数组的迭代解法

// dp[i] = x 表示， 当目标金额为i时， 至少需要x枚硬币

int coinChange(vector<int>& coins, int amount) {

  // 数组大小为amount + 1, 初始值为amount + 1
  vector<int> dp(amount + 1, amount + 1);
  
  //base case
  dp[0] = 0;
  for (int i = 0; i < dp.size(); i++) {
    // 内层for 在求所有子问题 + 1的 最小值
    for (int coin: coins) {
      //子问题无解， 跳过
      if (i - coin < 0) continue;
      dp[i] = min(dp[i], 1 + dp[i - coin]);
    }
  }
  return (dp[amount] == amount + 1) ? -1: dp[amount];
}


















